### PR TITLE
Improve registry chains testing with sandbox

### PR DESCRIPTION
**Pull request by Bolodya1997**
# Issues
1. Existing sandbox doesn't use registry client chain to register endpoints, so we can't use it for registry testing.
2. Existing sandbox doesn't allow to start a node without forwarder or with multiple forwarders, so we can't use it for testing interpose chain elements.
# Solution
Rework sandbox to the node API:
```go
// Node is a NSMgr with resources
type Node struct {
	NSMgr     *NSMgrEntry
	...
}

// NewForwarder starts a new forwarder and registers it on the node NSMgr
func (n *Node) NewForwarder(
	ctx context.Context,
	nse *registry.NetworkServiceEndpoint,
	generatorFunc token.GeneratorFunc,
	additionalFunctionality ...networkservice.NetworkServiceServer,
) (*EndpointEntry, error) {
	...
}

// NewEndpoint starts a new endpoint and registers it on the node NSMgr
func (n *Node) NewEndpoint(
	ctx context.Context,
	nse *registry.NetworkServiceEndpoint,
	generatorFunc token.GeneratorFunc,
	additionalFunctionality ...networkservice.NetworkServiceServer,
) (*EndpointEntry, error) {
	...
}

// NewClient starts a new client and connects it to the node NSMgr
func (n *Node) NewClient(
	ctx context.Context,
	generatorFunc token.GeneratorFunc,
	additionalFunctionality ...networkservice.NetworkServiceClient,
) networkservice.NetworkServiceClient {
	...
}
```
Usage:
```go
domain := sandbox.NewBuilder(t).
	SetNodesCount(nodesCount).
	SetRegistryProxySupplier(nil).
	SetContext(ctx).
	Build()

for _, node := range domain.Nodes {
	_, err := node.NewForwarder(ctx, new(registry.NetworkServiceEndpoint), sandbox.GenerateTestToken)
	require.NoError(t, err)
}

_, err := node.NewEndpoint(ctx, &registry.NetworkServiceEndpoint{...}, sandbox.GenerateTestTokent, ...endpointChain)
require.NoError(t, err)

client := node.NewClient(ctx, sandbox.GenerateTestTokent, ...clientChain)

conn, err := client.Request(...)
require.NoError(t, err)
validateConnection(t, conn)

_, err := client.Close(ctx, conn)
require.NoError(t, err)
```

### CODE REVIEW

File: pkg/networkservice/chains/nsmgrproxy/server_test.go [hunk 1]
```diff
@@ -21,36 +21,45 @@ import (
...
+		SetNodesCount(1).
+		SetContext(ctx).
+		SetDNSResolver(dnsServer).
+		Build()
+
+	_, err := domain.Nodes[0].NewForwarder(ctx, new(registry.NetworkServiceEndpoint), sandbox.GenerateTestToken)
+	require.NoError(t, err)
+
+	return domain
+}
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgrproxy/server_test.go:47]: this looks a little bit complex than before.

**Comment by denis-tingaikin** [comment]
> Existing sandbox doesn't use registry client chain to register endpoints, so we can't use it for registry testing.

It is not fully correct. You are able to create a registry client using nsmgr url.

**Comment by denis-tingaikin** [comment]
> Existing sandbox doesn't allow to start a node without forwarder or with multiple forwarders, so we can't use it for testing interpose chain elements.

We can add this, but need to think about how to not add more complexity :)

**Comment by Bolodya1997** [comment]
> > Existing sandbox doesn't use registry client chain to register endpoints, so we can't use it for registry testing.
> 
> It is not fully correct. You are able to create a registry client using nsmgr url.

You are right, it is not making it possible - it _reworks general sandbox way to add endpoint_ to use registry chain :)

File: pkg/networkservice/chains/nsmgrproxy/server_test.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/nsmgrproxy/server_test.go]: Yep, it is. Fixed this with `.SetNodeSetup` and `defaultSetupNode`:
```go
// SetupNodeFunc setups each node on Builder.Build() stage
type SetupNodeFunc func(ctx context.Context, node *Node)

// SetNodeSetup replaces default node setup to custom function
func (b *Builder) SetNodeSetup(f SetupNodeFunc) *Builder {
	b.setupNode = f
	return b
}

func defaultSetupNode(t *testing.T) SetupNodeFunc {
	return func(ctx context.Context, node *Node) {
		_, err := node.NewForwarder(ctx, new(registryapi.NetworkServiceEndpoint), GenerateTestToken)
		require.NoError(t, err)
	}
}
```

File: pkg/tools/sandbox/node.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	additionalFunctionality ...networkservice.NetworkServiceServer,
+) (*EndpointEntry, error) {
+	return n.newEndpoint(ctx, nse, generatorFunc, false, additionalFunctionality...)
+}
+
+func (n *Node) newEndpoint(
+	ctx context.Context,
+	nse *registry.NetworkServiceEndpoint,
+	generatorFunc token.GeneratorFunc,
+	isForwarder bool,
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/node.go:93]:
```suggestion
	registryClient registry.NetworkServiceEndpointRegistryClient
```

File: pkg/tools/sandbox/registry_client.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/networkservicemesh/sdk/pkg/registry/core/chain"
+)
+
+type registryClient struct {
+	nsClient  registry.NetworkServiceRegistryClient
+	nseClient registry.NetworkServiceEndpointRegistryClient
+}
+
+// NewRegistryClient returns NSE registry client to register NSE to registry
+func NewRegistryClient(ctx context.Context, cc grpc.ClientConnInterface, isForwarder bool) registry.NetworkServiceEndpointRegistryClient {
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/registry_client.go:41]: Can we just create chains for registry clients?

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/tools/sandbox/node.go (hunk 2 above)
Comment (Bolodya1997) [reply pkg/tools/sandbox/node.go]: Nice idea, replaced with this.

File: pkg/tools/sandbox/registry_client.go (hunk 3 above)
Comment (Bolodya1997) [reply pkg/tools/sandbox/registry_client.go]: Yep, done.

**Review by denis-tingaikin — APPROVED** [review]
Looks good

File: pkg/registry/core/interdomain/interdomain_nse_test.go [hunk 4]
```diff
@@ -223,7 +223,7 @@ func TestInterdomainFloatingNetworkServiceEndpointRegistry(t *testing.T) {
 
 	domain3 := sandbox.NewBuilder(t).
 		SetNodesCount(0).
-		SetRegistrySupplier(func(context.Context, *url.URL, ...grpc.DialOption) registry2.Registry {
+		SetRegistrySupplier(func(context.Context, time.Duration, *url.URL, ...grpc.DialOption) registry2.Registry {
```
Comment (denis-tingaikin) [inline pkg/registry/core/interdomain/interdomain_nse_test.go:226]: Can we fix the artifact `registry2` ? :)

Comment (Bolodya1997) [reply pkg/registry/core/interdomain/interdomain_nse_test.go]: Yep, fixed with `registryapi` for `github.com/networkservicemesh/api/pkg/api/registry` and no alias for `github.com/networkservicemesh/sdk/pkg/registry`.

**Pull request merged**